### ISSUE TITLE
task_ai_model : Investigating GTP2 conversational engine 

### ISSUE THREAD

**Issue by nlauchande** [omdena/ptsd#8]

**Comment by elateifsara** [comment]
Hello, I have been working on this tutorial because it generates a conversation that you can interact with while defining the 'persona' of the bot (like some previous knowledge on that 'person': bot.
I am still struggling to understand how they train the model so that we can use our own data. If anyone has experience with GPT-2 and would like to help then here’s the link to the article: https://medium.com/huggingface/how-to-build-a-state-of-the-art-conversational-ai-with-transfer-learning-2d818ac26313
And here’s the link to the Colab notebook (if you wish to collaborate on it please let me know): 
https://colab.research.google.com/drive/1JATyJIFTTjgZlRJLn18xF77zJIJXFWYJ

**Comment by Aya-S** [comment]
The tutorial I'm investigating is the one here: https://github.com/ilopezfr/gpt-2
It seems simpler so I will continue with experimenting with it.
PS: I sometimes have trouble opening medium articles due to ISP issues.

**Comment by albertlai431** [comment]
@elateifsara How's the chatbot going? I could collaborate with you on it
@Aya-S I believe the GPT-2 is used to generate fake articles, so the original github repo wouldn't be good for a chatbot. Good to play around with it though!

**Comment by Aya-S** [comment]
@albertlai431  there is a "Q&A" mode explored in the repo I linked above.
in gpt-2-playground_.ipynb. so seems there is some flexibility there.

**Comment by elateifsara** [comment]
@albertlai431 still training it on their data. I tried the interpret.py module for now as their train.py has some problems I think they are using some deprecated module from tensorlfow. 
![image](https://user-images.githubusercontent.com/27445092/61322369-6556c900-a805-11e9-922b-0ac79611993d.png)

**Comment by albertlai431** [comment]
@elateifsara that's strange, what sucks about NLP models is that they're super difficult to train 😅
Are you saying that we could explore converting our data into JSON format?

**Comment by elateifsara** [comment]
Yeah, @albertlai431 or even some column format for easy consumption and analysis. I think this should be our 2nd pressing issue.

**Comment by Aya-S** [comment]
An example I just ran with a  prompt I just  made up xD
I'm not specifying the length of the response so it goes on and on.

Model prompt >>> I have suffered from a traumatic incident.I was living with my family in the countryside. a three story house. my grandparents lived on the ground floor. my parents lived on the first floor. and my wife and I lived on the second floor. the third floor was empty. they has been a bombing nearby and so my whole family fled and we are now homeless. we are suffering. we have no shelter and my grandparents are very old. Q: can you help us ? A:

======================================== SAMPLE 1 ========================================
 yes, we will give money to their relatives and also we have a few cars that we can rent out. we can take them from the house towards the church, for example. if we can send them there, then they can go and help us. we have about four people who are living there. my son is in the hospital and he was killed by the shelling. we have about ten people who were in the hospital who are still alive and are doing well. the hospital is now open and our doctors are helping us and we are giving food to the wounded. we have a lot of volunteers coming as we speak. there are many wounded and dead inside the hospital but we are still trying to bring them back.. we are trying to help them, which is not easy. Q: so they are doing well ? A: yes, we are all very grateful. Q: thank you so much for your efforts. A: yes, I am very grateful. Q: and what have you given to the wounded ? A: we have given food and medicines to some people. that is about it, if you would like to write anything, please send it to us. thank you very much a very great effort !<|endoftext|>What is a 'Leverage'

Leverage is a market technique in which a company or business invests capital in an attempt at expanding its competitive advantage. The company uses its resources effectively to produce a profit and generate new customer orders.

BREAKING DOWN 'Leverage'

The use of leverage is common with a range of industries. A manufacturing firm could make use of the expertise it acquired in manufacturing to build on those skills to produce a product that is better at what it is selling. An information technology firm has the ability to obtain new customers with a simple website that is designed for them to use while developing new technology.

The use of leverage requires a certain amount of investment that is then spent on an investment in new products, services, or processes. An example could be using consultants to help a company decide how to improve a product or process and then investing in that process.

In order to use the leverage employed by a firm, the firm must be able to use its resources effectively to generate growth. This is what the firm must do to succeed and a firm that invests more capital into itself doesn't want to fail.

Leverage as a Technique

The use of leverage is a market technique. The purpose of using leverage is to create leverage so that a
======================================== SAMPLE 2 ========================================
 yes I can.Q: do you want to come with your family ? A: I could just go from the village to the hospital in the town. Q: so you are helping us to go to the hospital ? A: yes, i do want to see that this will work .but we need to be sure that this will work. my father is elderly . he used to work and has no job, just with us in the house. it makes it very difficult for us to find work .we do not have enough money .we are now homeless, we are afraid if there is more attacks , there may be more suicides and more accidents in the village .that I am afraid is what will happen .i am afraid for my family .they are afraid because the village people do not fear the attack because they know they can get help from the government . they do not fear when the government is not there .they do not fear anything. they feel scared and the villages fear it .Q: so that is why you want to come with your family ? A: yes .Q: the reason why the government did not go to help the victims ? A: we were not allowed to go to the hospital .because they did not know a village .and there was one victim that lived in the second floor of my grandmother's house and we were not allowed to visit there ,and the attack happened on the day of our anniversary ,our birthday .it was in the afternoon and it was a quiet day in the town ,but then suddenly it started bombing . and my grandmother was working . and then she heard on the news there were three attacks in the village .her husband was working in a plant and he knew his house was in danger from the bombing .so he went to the hospital and the police officers were there .they told him to go back to his house .he ran away from home .and his friend was waiting for him .the police did not come because they did not know if they would be allowed to visit .i was crying every day and I wanted to die ,like any normal person .and the police did not come .the next day the government did arrive .the villagers were waiting for the government and they told them to leave the village .because they did not know if they would be allowed .but they do not know how the villagers are affected .the government gave the villagers permission to go to the hospital .but they cannot leave the town because the village is not in their hands .they are afraid .they do not know
================================================================================

**Comment by elateifsara** [comment]
@Aya-S  I have one question is the 'leverage' thing included in the training? Or is it a technique used in NLP?

**Comment by Aya-S** [comment]
Experiment 2: limiting the answer to 20 tokens and providing some " context" trying to mimic what @elateifsara  was trying out. the first generated response is such a typical therapist response xD
Prompt: I know you are an excellent psychiatrist and you understand human psychology. I learned that you are helpful and can give helpful suggestions. I have suffered from a traumatic incident .I was living with my family in the countryside . a three story house. my grandparents lived on the ground floor. my parents lived on the first floor. and my wife and I lived on the second floor. the third floor was empty. they has been a bombing nearby and so my whole family fled and we are now homeless. we are suffering. we have no shelter and my grandparents are very old. 
Q:

**SAMPLE1:**
 If I ask you to give me some advice, what would you tell me? A: I would tell you to go abroad to live with friends. if we have to live in a shelter and
**SAMPLE 2:** 
 How do the relatives help the family ? Q:  We pray for the families to give them food and we pray that our elderly relatives be allowed to remain in their homes. our children are

**Comment by Aya-S** [comment]
> @Aya-S I have one question is the 'leverage' thing included in the training? Or is it a technique used in NLP?

sorry just saw this. the "leverage" part  is something completely irrelevant, in my opinion, the coherence isn't very good

**Comment by albertlai431** [comment]
@Aya-S Haha, that's pretty funny, and it's definitely a start 😂
Is that with Open-AI's GPT-2 turned chatbot?

**Comment by elateifsara** [comment]
Thank you @Aya-S if the team is interested we can keep improving it.

**Comment by elateifsara** [comment]
@albertlai431 and @Aya-S did you get how gpt2 works I mean I think we can the tutorial I have been working on and [this therapy examples](https://www.apa.org/ptsd-guideline/resources/case-examples) to generate data: Because they have a background story == 'persona' in the tutorial I worked on and there's some kind of small dialog==CBT. I do believe it will provide us good results and would be a first good work towards CBT.

**Comment by albertlai431** [comment]
@elateifsara Yeah, I think it would be a good try! However, I think for task 6 and 7, our most pressing tasks right now, we're focusing on logistic regression + LSTM? Would this be correlated, or would this be a longer term backend project?

**Comment by elateifsara** [comment]
Hy @albertlai431 yeah, you’re right. I am thinking about exploring the use of CNN (it gave good results before in one of my NLP classification projects). We should try as many promising models as we can until we get high results.
